src/vector_of_array.jl

```diff
@@ -707,30 +717,37 @@ end
 
 # Tools for creating similar objects
 Base.eltype(::Type{<:AbstractVectorOfArray{T}}) where {T} = T
-# TODO: Is there a better way to do this?
+
 @inline function Base.similar(VA::AbstractVectorOfArray, args...)
     if args[end] isa Type
-        return Base.similar(eltype(VA)[], args..., size(VA))
+        return Base.similar(VA.u, args..., size(VA))
     else
-        return Base.similar(eltype(VA)[], args...)
+        return Base.similar(VA.u, args...)
     end
 end
-@inline function Base.similar(VA::VectorOfArray, ::Type{T} = eltype(VA)) where {T}
-    VectorOfArray([similar(VA[:, i], T) for i in eachindex(VA.u)])
-end
 
-# for VectorOfArray with multi-dimensional parent arrays of arrays where all elements are the same type
 function Base.similar(vec::VectorOfArray{
         T, N, AT}) where {T, N, AT <: AbstractArray{<:AbstractArray{T}}}
     return VectorOfArray(similar.(Base.parent(vec)))
 end
 
-# special-case when the multi-dimensional parent array is just an AbstractVector (call the old method)
-function Base.similar(vec::VectorOfArray{
-        T, N, AT}) where {T, N, AT <: AbstractVector{<:AbstractArray{T}}}
-    return Base.similar(vec, eltype(vec))
+@inline function Base.similar(VA::VectorOfArray, ::Type{T} = eltype(VA)) where {T}
+    VectorOfArray([similar(VA[:, i], T) for i in eachindex(VA.u)])
 end
 
+@inline function Base.similar(VA::VectorOfArray, dims::N) where {N <: Number}
+    l = length(VA)
+    if dims <= l
+        VectorOfArray([similar(VA[:, i]) for i in 1:l])
```

Comment:
This assumes the internal is a vector? Why not `VectorOfArray(similar.(VA.u))`?

Reply:
Yeah, broadcast is better